File: java_concurrent_testing/src/main/java/com/example/test/QueueConsistencyTest.java
package com.example.test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import com.example.utils.QueueADT;

public class QueueConsistencyTest {
    private static int numberOfThreads = 5;
    private static int maxRunningTime = 2;
    private static int keyRange = 100;

    public static void main(String[] args) {
        try {
            for (int i = 0; i < args.length - 1; i++) {
                if ("-n".equals(args[i])) numberOfThreads = Integer.parseInt(args[i + 1]);
                if ("-d".equals(args[i])) maxRunningTime = Integer.parseInt(args[i + 1]);
                if ("-k".equals(args[i])) keyRange = Integer.parseInt(args[i + 1]);
            }
            QueueADT queue;
            try {
                queue = (QueueADT) Class.forName("com.example.Sets.ConcurrentDataStructure")
                        .getDeclaredConstructor().newInstance();
            } catch (Exception e) {
                System.out.println("Sanity Test Failed");
                return;
            }

            // === Phase 1: Concurrent conservation check ===
            AtomicInteger totalEnqueued = new AtomicInteger(0);
            AtomicInteger totalDequeued = new AtomicInteger(0);
            AtomicInteger unused = new AtomicInteger(0);
            AtomicLong orderCounter = new AtomicLong(0);
            int[] results = new int[numberOfThreads];
            List<List<Integer>> allDequeued = new ArrayList<>(Collections.nCopies(numberOfThreads, null));
            for (int i = 0; i < numberOfThreads; i++) allDequeued.set(i, new ArrayList<>());

            Thread[] threads = new Thread[numberOfThreads];
            for (int i = 0; i < numberOfThreads; i++) {
                threads[i] = new Thread(new RunOperationsQueue(
                        queue, i, keyRange, results,
                        totalEnqueued, totalDequeued, unused, true,
                        orderCounter, allDequeued));
            }
            RunController.startFlag = RunController.stopFlag = false;
            for (Thread t : threads) t.start();
            RunController.startFlag = true;
            Thread.sleep(maxRunningTime * 1000L);
            RunController.stopFlag = true;
            for (Thread t : threads) t.join();

            int remaining = 0;
            while (!queue.isEmpty()) {
                int v = queue.dequeue();
                if (v == -1) break;
                remaining++;
            }
            int enq = totalEnqueued.get();
            int deq = totalDequeued.get();
            boolean conservationOk = (enq >= deq) && (enq <= deq + remaining + numberOfThreads * 2);

            // === Phase 2: Per-Producer FIFO ordering check ===
            // Each producer thread enqueues values encoded as (threadId * OFFSET + seqNum).
            // Multiple consumer threads dequeue concurrently.
            // Then we verify: for each producer, its items appear in the global
            // dequeue order with strictly increasing sequence numbers.
            // This is the correct concurrent FIFO invariant.
            final int OFFSET = 1_000_000;
            final int ITEMS_PER_PRODUCER = 500;
            final int NUM_PRODUCERS = numberOfThreads;

            QueueADT queue2;
            try {
                queue2 = (QueueADT) Class.forName("com.example.Sets.ConcurrentDataStructure")
                        .getDeclaredConstructor().newInstance();
            } catch (Exception e) {
                System.out.println("Sanity Test Failed");
                return;
            }

            // Phase 2a: All producers enqueue concurrently
            CyclicBarrier enqBarrier = new CyclicBarrier(NUM_PRODUCERS);
            Thread[] producers = new Thread[NUM_PRODUCERS];
            for (int t = 0; t < NUM_PRODUCERS; t++) {
                final int tid = t;
                producers[t] = new Thread(() -> {
                    try { enqBarrier.await(); } catch (Exception ignored) {}
                    for (int s = 0; s < ITEMS_PER_PRODUCER; s++) {
                        queue2.enqueue(tid * OFFSET + s);
                    }
                });
            }
            for (Thread t : producers) t.start();
            for (Thread t : producers) t.join();

            // Phase 2b: Single-threaded dequeue to get deterministic ordering
            // The concurrent correctness was exercised in the enqueue phase above.
            // Using a single dequeue thread avoids the race between dequeue() and
            // list.add() that corrupts the global order with concurrent consumers.
            List<Integer> globalDequeueOrder = new ArrayList<>();
            while (true) {
                int v = queue2.dequeue();
                if (v == -1) break;
                globalDequeueOrder.add(v);
            }

            // Phase 2c: Verify per-producer ordering
            // For each producer, its items must appear in increasing seqNum order
            boolean fifoOk = true;
            String fifoDetail = "";
            for (int prod = 0; prod < NUM_PRODUCERS; prod++) {
                int lastSeq = -1;
                for (int val : globalDequeueOrder) {
                    int origProducer = val / OFFSET;
                    int seqNum = val % OFFSET;
                    if (origProducer == prod) {
                        if (seqNum <= lastSeq) {
                            fifoDetail = String.format(
                                "producer %d: seq %d appeared after seq %d", prod, seqNum, lastSeq);
                            fifoOk = false;
                            break;
                        }
                        lastSeq = seqNum;
                    }
                }
                if (!fifoOk) break;
            }

            int totalExpected = NUM_PRODUCERS * ITEMS_PER_PRODUCER;
            int totalGot = globalDequeueOrder.size();

            boolean passed = conservationOk && fifoOk;

            if (passed) {
                System.out.println("Enqueued: " + enq + ", Dequeued: " + deq + ", Remaining: " + remaining);
                System.out.println("FIFO per-producer check: " + NUM_PRODUCERS + " producers x "
                        + ITEMS_PER_PRODUCER + " items, " + totalGot + "/" + totalExpected + " dequeued OK");
                System.out.println("Sanity Test Passed");
            } else {
                if (!conservationOk)
                    System.out.printf("Queue conservation failed: enqueued=%d dequeued=%d remaining=%d%n", enq, deq, remaining);
                if (!fifoOk)
                    System.out.println("Queue FIFO ordering failed: " + fifoDetail);
                System.out.println("Sanity Test Failed");
            }
        } catch (Throwable t) {
            System.out.println("Sanity Test Failed");
        }
    }
}

File: cpp_concurrent_testing/test/QueueConsistencyTest.cpp
#include "TestConfig.hpp"
#include "RunOperationsQueue.hpp"
#include "RunController.hpp"
#include <iostream>
#include <vector>
#include <thread>
#include <memory>
#include <chrono>
#include <atomic>
#include <functional>

#include "StructureFactory.hpp"

/**
 * Queue Consistency Test — mirrors Java QueueConsistencyTest.
 *
 * Phase 1: Concurrent conservation check
 *   Threads enqueue/dequeue concurrently. Afterwards:
 *     enqueued >= dequeued  AND  enqueued <= dequeued + remaining + margin
 *
 * Phase 2: FIFO per-producer ordering check
 *   Each producer encodes values as (tid * OFFSET + seqNum).
 *   After all producers finish, a single-threaded drain verifies that
 *   for each producer, its items appear in strictly increasing seqNum order.
 */
int main(int argc, char** argv) {
    TestConfig config;
    if (!config.parse(argc, argv)) return 1;

    auto queue = createQueue(config.algo);
    if (!queue) {
        std::cerr << "Algorithm " << config.algo << " not imp.\n";
        return 1;
    }

    std::cout << "Running Consistency Test for Queue: " << config.algo << "\n";

    // ── Phase 1: Concurrent conservation check ──────────────────────────
    std::vector<int> results(config.numThreads, 0);
    std::vector<std::thread> threads;
    std::atomic<int> totalEnqueued{0}, totalDequeued{0};
    RunController::reset();

    for (int i = 0; i < config.numThreads; i++) {
        threads.emplace_back(RunOperationsQueue::sanityRun, queue.get(), i,
                             config.keyspaceSize, std::ref(results),
                             std::ref(totalEnqueued), std::ref(totalDequeued));
    }

    RunController::startFlag.store(true);
    std::this_thread::sleep_for(std::chrono::seconds(config.duration));
    RunController::stopFlag.store(true);

    for (auto& t : threads) t.join();

    // Drain remaining items
    int remaining = 0;
    while (!queue->isEmpty()) {
        int v = queue->dequeue();
        if (v == -1) break;
        remaining++;
    }

    int enq = totalEnqueued.load();
    int deq = totalDequeued.load();

    // Identity: enq == deq + remaining (within a small margin for concurrent dequeues)
    bool conservationOk = (enq >= deq) && (enq <= deq + remaining + config.numThreads * 2);

    // ── Phase 2: Per-Producer FIFO ordering check ───────────────────────
    const int OFFSET = 1000000;
    const int ITEMS_PER_PRODUCER = 500;
    const int NUM_PRODUCERS = config.numThreads;

    auto queue2 = createQueue(config.algo);
    if (!queue2) {
        std::cout << "Sanity Test Failed\n";
        return 1;
    }

    // Phase 2a: All producers enqueue concurrently
    std::vector<std::thread> producers;
    std::atomic<bool> goFlag{false};

    for (int t = 0; t < NUM_PRODUCERS; t++) {
        producers.emplace_back([&, t]() {
            while (!goFlag.load(std::memory_order_acquire))
                std::this_thread::yield();
            for (int s = 0; s < ITEMS_PER_PRODUCER; s++) {
                queue2->enqueue(t * OFFSET + s);
            }
        });
    }
    goFlag.store(true, std::memory_order_release);
    for (auto& t : producers) t.join();

    // Phase 2b: Single-threaded dequeue to get deterministic ordering
    std::vector<int> globalDequeueOrder;
    while (true) {
        int v = queue2->dequeue();
        if (v == -1) break;
        globalDequeueOrder.push_back(v);
    }

    // Phase 2c: Verify per-producer ordering
    bool fifoOk = true;
    std::string fifoDetail;
    for (int prod = 0; prod < NUM_PRODUCERS; prod++) {
        int lastSeq = -1;
        for (int val : globalDequeueOrder) {
            int origProducer = val / OFFSET;
            int seqNum = val % OFFSET;
            if (origProducer == prod) {
                if (seqNum <= lastSeq) {
                    fifoDetail = "producer " + std::to_string(prod) +
                                 ": seq " + std::to_string(seqNum) +
                                 " appeared after seq " + std::to_string(lastSeq);
                    fifoOk = false;
                    break;
                }
                lastSeq = seqNum;
            }
        }
        if (!fifoOk) break;
    }

    int totalExpected = NUM_PRODUCERS * ITEMS_PER_PRODUCER;
    int totalGot = (int)globalDequeueOrder.size();

    bool passed = conservationOk && fifoOk;

    if (passed) {
        std::cout << "Enqueued: " << enq << ", Dequeued: " << deq << ", Remaining: " << remaining << "\n";
        std::cout << "FIFO per-producer check: " << NUM_PRODUCERS << " producers x "
                  << ITEMS_PER_PRODUCER << " items, " << totalGot << "/" << totalExpected << " dequeued OK\n";
        std::cout << "Sanity Test Passed\n";
        return 0;
    } else {
        if (!conservationOk)
            std::cout << "Queue conservation failed: enqueued=" << enq
                      << " dequeued=" << deq << " remaining=" << remaining << "\n";
        if (!fifoOk)
            std::cout << "Queue FIFO ordering failed: " << fifoDetail << "\n";
        std::cout << "Sanity Test Failed\n";
        return 1;
    }
}
